### PR TITLE
Implement curTokenIs, nextTokenIs and test cases

### LINKED ISSUE #16: [parser] Implement curTokenIs
implement `curTokenIs` **w/ test cases**

take token from TokenBuffer and check token type with its `t` input parameters

### PR DESCRIPTION
**Pull request by boohyunsik**
resolved: #16, #17, #18, #19 

details: Implement curTokenIs, nextTokenIs and test cases

- [x] Test case

### CODE REVIEW

File: parse/parser_internal_test.go [hunk 1]
```diff
@@ -1,7 +1,86 @@
...
+func (m *MockTokenBuffer) Read() Token {
+	ret := m.buf[m.sp]
+	m.sp++
+	return ret
+}
+
+func (m *MockTokenBuffer) Peek(n peekNumber) Token {
+	return m.buf[n]
+}
+
```
Comment (junbeomlee) [inline parse/parser_internal_test.go:37]: I think
`return m.buf[n]` should be `return m.buf[m.sp+n]`

File: parse/parser_internal_test.go [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+func TestParse_curTokenIs(t *testing.T) {
+	test := MockTokenBuffer{
+		buf: []Token{
```
Comment (junbeomlee) [inline parse/parser_internal_test.go:40]: It would be better to name the variable tokenBuf not test.
`tokenBuf := `

**Comment by hihiboss** [comment]
LGTM!

But... I think that it would be better if precedence tests can cover more cases (Now, they just test only LOWEST...)

**Review by hihiboss — APPROVED** [review]

File: parse/parser.go [hunk 3]
```diff
@@ -24,9 +24,19 @@ type precedence int
 
 const (
 	_ precedence = iota
+	LOWEST
+	EQUALS      // ==
+	LESSGREATER // > or <
+	SUM         // +
+	PRODUCT     // *
+	PREFIX      // -X or !X
+	CALL        // function(X)
```
Comment (zeroFruit) [inline parse/parser.go:31]: I think we don't have function now, so the `CALL` is not need for now

File: parse/parser_internal_test.go [hunk 4]
```diff
@@ -1,7 +1,67 @@
...
+	m.sp++
+	return ret
+}
+
+func (m *MockTokenBuffer) Peek(n peekNumber) Token {
+	return m.buf[m.sp+int(n)]
+}
+
+func TestParse_curTokenIs_nextTokenIs(t *testing.T) {
+	tokenBuf := MockTokenBuffer{
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: What about separating curTokenIs, nextTokenIs test case?

File: parse/parser_internal_test.go [hunk 5]
```diff
@@ -32,3 +92,32 @@ func TestPeekNumber_IsValid(t *testing.T) {
 		}
 	}
 }
+
+func TestParse_precedence(t *testing.T) {
+	tokenBuf := MockTokenBuffer{
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: separating curPrecedence, nextPrecedence might be good

**Review by junbeomlee — CHANGES_REQUESTED** [review]
In the book, precedence check indicates the rank used in expression calculation. It seems that int and Ident should not exist.

```
const (
	_ int = iota
	LOWEST
	EQUALS      // ==
	LESSGREATER // > or <
	SUM         // +
	PRODUCT     // *
	PREFIX      // -X or !X
	CALL        // myFunction(X)
)

var precedences = map[token.TokenType]int{
	token.EQ:       EQUALS,
	token.NOT_EQ:   EQUALS,
	token.LT:       LESSGREATER,
	token.GT:       LESSGREATER,
	token.PLUS:     SUM,
	token.MINUS:    SUM,
	token.SLASH:    PRODUCT,
	token.ASTERISK: PRODUCT,
	token.LPAREN:   CALL,
}
```

File: parse/parser_internal_test.go (hunk 4 above)
Comment (boohyunsik) [reply parse/parser_internal_test.go]: First time I separated these two test case functions, but in that case there are duplicated code like making tokenBuf code. So I combined curTokenIs and nextTokenIs test case. But, how about other people's think?

File: parse/parser_internal_test.go (hunk 1 above)
Comment (boohyunsik) [reply parse/parser_internal_test.go:37]: Ok, I modified it.

File: parse/parser_internal_test.go (hunk 2 above)
Comment (boohyunsik) [reply parse/parser_internal_test.go:40]: Also modified.

**Comment by boohyunsik** [comment]
> In the book, precedence check indicates the rank used in expression calculation. It seems that int and Ident should not exist.
> 
> ```
> const (
> 	_ int = iota
> 	LOWEST
> 	EQUALS      // ==
> 	LESSGREATER // > or <
> 	SUM         // +
> 	PRODUCT     // *
> 	PREFIX      // -X or !X
> 	CALL        // myFunction(X)
> )
> 
> var precedences = map[token.TokenType]int{
> 	token.EQ:       EQUALS,
> 	token.NOT_EQ:   EQUALS,
> 	token.LT:       LESSGREATER,
> 	token.GT:       LESSGREATER,
> 	token.PLUS:     SUM,
> 	token.MINUS:    SUM,
> 	token.SLASH:    PRODUCT,
> 	token.ASTERISK: PRODUCT,
> 	token.LPAREN:   CALL,
> }
> ```



> LGTM!
> 
> But... I think that it would be better if precedence tests can cover more cases (Now, they just test only LOWEST...)

Ok, I will modify it.

**Comment by boohyunsik** [comment]
> LGTM!
> 
> But... I think that it would be better if precedence tests can cover more cases (Now, they just test only LOWEST...)

Ok, I will think about it but actually, We can implement test cases after 'precedences' value be fixed.
I think about that first.

File: parse/parser_internal_test.go (hunk 4 above)
Comment (hihiboss) [reply parse/parser_internal_test.go]: Then, how about making a function like `setupMockTokenBuffer() TokenBuffer`, `buildMockTokenBuffer() TokenBuffer` in test file. And just call and sign it to `tokenBuf`!

```go
func TestParse_curTokenIs(t *testing.T) {
    tokenBuf := setupMockTokenBuffer()
    ret := curTokenIs(&tokenBuf, Int)
    ...
}
```

Like this!

**Comment by hihiboss** [comment]
@boohyunsik 
Ok, I understood!

File: parse/parser_internal_test.go (hunk 4 above)
Comment (boohyunsik) [reply parse/parser_internal_test.go]: Right. I will do it.

File: parse/parser.go (hunk 3 above)
Comment (boohyunsik) [reply parse/parser.go:31]: OK. It was from our book. I need to think more about it.

**Comment by boohyunsik** [comment]
I modified some codes.
1. Divided test cases to cur and next.
2. Set temporarily precedence map.

File: parse/parser.go (hunk 3 above)
Comment (hihiboss) [reply parse/parser.go:31]: How did it go “CALL”?? Need it or not?

**Comment by zeroFruit** [comment]
plz remove last commit message

> Merge branch 'develop' into develop

resolve conflict by rebasing the `develop` branch from your branch

**Comment by boohyunsik** [comment]
> plz remove last commit message
> 
> > Merge branch 'develop' into develop
> 
> resolve conflict by rebasing the `develop` branch from your branch

Yeap. I just have rebased but commit messasge was automatically made ;-(..
I modifed commit message. Please review it~

**Comment by zeroFruit** [comment]
LGTM 👍

**Review by zeroFruit — APPROVED** [review]

**Review by junbeomlee — APPROVED** [review]

**Comment by zeroFruit** [comment]
plz rebase again and resolve conflict

**Comment by boohyunsik** [comment]
> plz rebase again and resolve conflict

I did it! :-D

**Pull request merged**